bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+        pkg: LearnedSharePkg,
+    },
+}
+
+impl State {
+    pub fn name(&self) -> &'static str {
+        match self {
+            State::Uninitialized => "uninitialized",
+            State::InitialMember { .. } => "initial_member",
+            State::Learning => "learning",
+            State::Learned { .. } => "learned",
+        }
+    }
+}
+
+/// A response to an Fsm API request
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ApiOutput {
+    /// This peer has been initialized
+    ///
+    /// The caller *must* persist the state
+    PeerInitialized,
+
+    /// Rack initialization has completed. This node was the coordinator.
+    RackInitComplete,
+
+    /// A `RackSecret` was reconstructed
+    RackSecret { request_id: Uuid, secret: RackSecret },
+
+    /// An extra share has been distributed to a learning peer
+    ///
+    /// The caller *must* persist the state
+    ShareDistributedToLearner,
+
+    /// This peer Learned its share
+    ///
+    /// The caller *must* persist the state
+    LearningCompleted,
+}
+
+/// An error returned from an Fsm API request
+#[derive(Error, Debug, Clone, PartialEq, Eq)]
+pub enum ApiError {
+    #[error("already initialized")]
+    AlreadyInitialized,
+
+    #[error("not yet initialized")]
+    NotInitialized,
+
+    #[error("cannot retrieve or distribute shares while learning")]
+    StillLearning,
+
+    #[error("rack init timeout: unacked_peers: {unacked_peers:?}")]
+    RackInitTimeout { unacked_peers: BTreeSet<Baseboard> },
+
+    #[error("rack init falied: trust quorum error: {0:?}")]
+    RackInitFailed(TrustQuorumError),
+
+    #[error("rack secret load timeout")]
+    RackSecretLoadTimeout,
+
+    #[error("share from {from} has invalid sha3_256 digest")]
+    InvalidShare { from: Baseboard },
+
+    #[error("critical: failed to reconstruct rack secret with valid shares")]
+    FailedToReconstructRackSecret,
+
+    #[error("critical: failed to decrypt extra shares")]
+    FailedToDecryptExtraShares,
+
+    #[error("unexpected response ({msg}) from ({from}) in state ({state}) with request_id ({request_id})")]
+    UnexpectedResponse {
+        from: Baseboard,
+        state: &'static str,
+        request_id: Uuid,
+        msg: &'static str,
+    },
+
+    #[error("error response received from ({from}) in state ({state}) with request_id ({request_id}): {error:?}")]
+    ErrorResponseReceived {
+        from: Baseboard,
+        state: &'static str,
+        request_id: Uuid,
+        error: MsgError,
+    },
+}
+
+pub struct Fsm {
+    /// The current state of this peer
+    state: State,
+    /// Unique ID of this peer
+    id: Baseboard,
+
+    /// User provided configuration
+    config: Config,
+
+    /// Unique IDs of connected peers
+    connected_peers: BTreeSet<Baseboard>,
+
+    /// Manage all trackable requests
+    request_manager: RequestManager,
+
+    /// Envelopes not managed by the `RequestManager`
+    /// These are all envelopes containing `Response` messages
```

Comment:
If these are all responses, then maybe name this member `responses` instead of `envelopes`.

Reply:
agreed!